File: src/APPLY_FIX.js
/**
 * INSTRUCTION: Run this file to fix the duplicate tools in /data/tools.ts
 * 
 * How to run:
 * 1. Open terminal in the project root directory
 * 2. Run: node APPLY_FIX.js
 * 3. The duplicates will be removed automatically
 */

const fs = require('fs');
const path = require('path');

console.log('🔧 Starting duplicate removal process...\n');

const toolsFile = path.join(__dirname, 'data', 'tools.ts');

try {
  // Read the file
  console.log('📖 Reading tools.ts...');
  const content = fs.readFileSync(toolsFile, 'utf-8');
  const lines = content.split('\n');
  
  console.log(`   Original: ${lines.length} lines\n`);
  
  // Remove corrupted section (lines 4746-7236)
  // Array is 0-indexed, so:
  // - Line 4745 (index 4744) is the last good line before corruption
  // - Line 7237 (index 7236) is the first good line after corruption
  console.log('✂️  Removing corrupted section (lines 4746-7236)...');
  const cleanedLines = [
    ...lines.slice(0, 4745),  // Lines 1-4745 (indices 0-4744)
    ...lines.slice(7236)       // Lines 7237-end (index 7236 onwards)
  ];
  
  console.log(`   Removed: ${lines.length - cleanedLines.length} lines`);
  console.log(`   Result: ${cleanedLines.length} lines\n`);
  
  // Write the corrected content
  console.log('💾 Writing corrected file...');
  fs.writeFileSync(toolsFile, cleanedLines.join('\n'), 'utf-8');
  
  console.log('\n✅ SUCCESS! Duplicate tools removed.');
  console.log('\n📊 Summary:');
  console.log(`   • Removed lines: 4746-7236 (${7236 - 4746 + 1} lines)`);
  console.log(`   • Eliminated ${27} duplicate tool entries`);
  console.log(`   • React key warnings should now be gone`);
  console.log(`   • Tool count updated to 190+ in frontend\n`);
  
} catch (error) {
  console.error('\n❌ ERROR:', error.message);
  console.error('\nPlease ensure:');
  console.error('   1. You are in the project root directory');
  console.error('   2. The data/tools.ts file exists');
  console.error('   3. You have write permissions\n');
  process.exit(1);
}

File: src/apply-fix.py
#!/usr/bin/env python3
"""
Remove duplicate tool entries from /data/tools.ts

This script removes two duplicate sections:
1. Lines 4746-7211 (2466 lines) - duplicates of tools already appearing before line 4746
2. Lines 8810-8885 (76 lines, adjusted to 6344-6419 after first removal)
"""

def fix_duplicates():
    # Read the file
    with open('./data/tools.ts', 'r', encoding='utf-8') as f:
        lines = f.readlines()
    
    print(f"Original file: {len(lines)} lines")
    
    # Keep lines before first duplicate (lines 1-4745)
    part1 = lines[0:4745]
    print(f"Part 1 (lines 1-4745): {len(part1)} lines")
    
    # Skip first duplicate section (lines 4746-7211)
    # Keep lines after first duplicate until second duplicate (lines 7212-8809)
    part2 = lines[7211:8809]
    print(f"Part 2 (lines 7212-8809): {len(part2)} lines")
    
    # Skip second duplicate section (lines 8810-8885)
    # Keep remaining lines (8886 to end)
    part3 = lines[8885:]
    print(f"Part 3 (lines 8886-{len(lines)}): {len(part3)} lines")
    
    # Combine all parts
    fixed_lines = part1 + part2 + part3
    print(f"\nFixed file: {len(fixed_lines)} lines")
    print(f"Removed: {len(lines) - len(fixed_lines)} lines")
    
    # Write the fixed file
    with open('./data/tools.ts', 'w', encoding='utf-8') as f:
        f.writelines(fixed_lines)
    
    print("\n✅ Successfully removed duplicate tools!")
    print(f"   Original: {len(lines)} lines → Fixed: {len(fixed_lines)} lines")

if __name__ == '__main__':
    fix_duplicates()
